src/main/java/racing/Strategy.java

```diff
@@ -0,0 +1,19 @@
+package racing;
+
+import java.util.Random;
+
+public class Strategy {
+    private static final int MAX_RANDOM_NUMBER = 10;
+    private static final int MOVE_THRESHOLD = 4;
+
+    public boolean isMovable = false;
+
+    public void setMovable(){
+        Random random = new Random();
+        isMovable = random.nextInt(MAX_RANDOM_NUMBER) >= MOVE_THRESHOLD;
+    }
+
+    public void setMovable(boolean move){
+        isMovable = move;
+    }
+}
```

Comment:
```suggestion
public class Strategy {
    private static final int MAX_RANDOM_NUMBER = 10;
    private static final int MOVE_THRESHOLD = 4;

    private final Random random = new Random();

    public boolean isMovable(){
        return random.nextInt(MAX_RANDOM_NUMBER) >= MOVE_THRESHOLD;
    }
}
```
`isMovable`만 열어두면 되지 않을까요?

Reply:
전략을 도메인 Car에게 전달되기는 하지만 아직 까지 테스트를 원하는대로 주무르지 못하고 있네요
그럼 결국에는 같은 현상입니다.

이럴 때는 Strategy라는 구현체 타입을 명시하는 것이 아닌 인터페이스 타입을 받도록 선언하면 됩니다.
```java
public interface Strategy { // 전략 인터페이스 
    boolean isMovable();
} 

//   -------------

public class RandomStrategy implements Strategy { // 전략 인터페이스의 구현체
    private static final int MAX_RANDOM_NUMBER = 10;
    private static final int MOVE_THRESHOLD = 4;
    
    private final Random random = new Random();
    
    @Override
    public boolean isMovable(){
        return random.nextInt(MAX_RANDOM_NUMBER) >= MOVE_THRESHOLD;
    }
}

// -------------
public void moveCar(Strategy strategy) { // 전략 인터페이스를 의존
    if (strategy.isMovable()) {
        moveCount += 1;
    }
}

// -------------
car.moveCar(new RandomStrategy()); // 랜덤 구현체 주입
```

Reply:
이렇게 변경 가능성이 있는 부분을 직접 의존하지 않고
인터페이스를 의존하면 의존 관계가 역전이 됩니다.

### 기존 방식 
실제 구현체에 의존하고 있기 때문에 변경이 어렵지만
Car --> Random 전략
Car --> 날씨 전략 (랜덤에서 변경이 어려움)

### 의존 관계 역전
인터페이스에 의존하고 있기 때문에 여러 구현체만 상황에 맞게 갈아 끼우기만 하면 됩니다.
- Car --> Strategy 인터페이스  <-- Random 전략
- Car --> Strategy 인터페이스  <-- 무조건 전진 전략 (변경 쉬움)
- Car --> Strategy 인터페이스  <-- 무조건 정지 전략 (변경 쉬움) 
- Car --> Strategy 인터페이스  <-- 실시간 날씨 전략  (변경 쉬움)

Reply:
이러면 테스트 코드에서 어떤 전략이든 쉽게 바꿔가면서 테스트가 가능할 것 같습니다.

DIP를 검색해 보시면 좋은 글들이 많으니 참고하시면 좋겠네요.

Reply:
회사에서도 패턴을 사용해 구현하려고하는데 익숙치 않아 정확한 구현이 잘 되지 않았는데 이번기회에 느리더라도 확실히 알고 넘어갈 수 있도록 하겠습니다! 

알려주셔서 너무 감사합니다!